### PR TITLE
Hpp challenge

### PR DESCRIPTION
**Pull request by shoebpate1**

### CODE REVIEW

File: routes/basketItems.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+const utils = require('../lib/utils')
+const challenges = require('../data/datacache').challenges
+const insecurity = require('../lib/insecurity')
+const models = require('../models/index')
+const clarinet = require('clarinet')
+
+module.exports = function addBasketItem () {
+  return (req, res, next) => {
+    const parser = clarinet.parser()
```
Comment (J12934) [inline routes/basketItems.js]: Can you extract the custom json parser out into a function?
Like "function ParseJsonCustom(jsonString)"?

File: routes/basketItems.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    const result = []
+    parser.onkey = parser.onopenobject = k => {
+      result.push({key: k, value: null})
+    }
+    parser.onvalue = v => {
+      result[result.length - 1].value = v
+    }
+    parser.write(req.rawBody).close()
+
+    var productIds = []
```
Comment (J12934) [inline routes/basketItems.js]: This code is for adding multiple products to the basket right?
Is that even possible in the Application?

I guess the code would be much leaner if the code would just take one product.

File: server.js [hunk 3]
```diff
@@ -121,7 +122,18 @@ app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument))
 
 app.use(express.static(applicationRoot + '/app'))
 app.use(cookieParser('kekse'))
-app.use(bodyParser.json())
+
+/* File Upload */
+app.post('/file-upload', upload.single('file'), fileUpload())
+
+app.use(bodyParser.text({type: '*/*'}))
+app.use(function jsonParser (req, res, next) {
```
Comment (J12934) [inline server.js:130]: This custom bodyParser.json() implementation is for saving the body into rawBody right?
I guess we could do it like this, to keep using the functionality provided by bodyParser.
```js
app.use(bodyParser.text({type: '*/*'}))
app.use((req, res, next) => {
    req.rawBody = req.body()
    next()
})
app.use(bodyParser.json())
```

**Review by J12934 — CHANGES_REQUESTED** [review]
Okay very cool that you found a way to do this this fast! 👍

But there are some issues in this PR:
1. The Frontend doesn't appear to be using the new custom endpoint. The UI sends a `PUT` request which gets handled by the epilogue (which is still active and should be taken out). Your new endpoint listens on POST Requests.
2. Some parts of the codes are a bit hard to read. I added some inline comments for these parts with suggestion how to improve the readability.

Hit me up on slack if you got any questions (faster than doing it on github)

File: routes/basketItems.js (hunk 2 above)
Comment (shoebpate1) [reply routes/basketItems.js]: No, @J12934 productId`s` is for storing multiple instancesof key which are in rawBody of POST, not for adding multiple products. It just takes the last value like JSON normally does. This POST endpoint exactly works like epilogue POST

File: server.js (hunk 3 above)
Comment (shoebpate1) [reply server.js:130]: I tried this but it doesnt work @J12934.

Comment (J12934) [reply server.js:130]: Just tried it there was a slight error in the code i submitted it has to be `req.body` not `req.body()`.
Other than that it seems to work fine.

Comment (shoebpate1) [reply server.js:130]: Oh can you confirm by running e2e test or something actually i did tried it with `req.body
`

Comment (J12934) [reply server.js:130]: Mh no your right tests are failing. I wonder why 🤔

Comment (shoebpate1) [reply server.js:130]: Actually after using it, json doesn't get parsed I guess .. I tried logging in and it wasn't working so I confirmed it doesn't work.

**Review by J12934 — APPROVED** [review]

**Pull request merged**